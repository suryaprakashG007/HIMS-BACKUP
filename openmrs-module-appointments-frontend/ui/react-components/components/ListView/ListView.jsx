import React, { useState, useEffect } from 'react'
import classNames from 'classnames';
import { appointmentListView, noAppointmentsListView, sortIcon, selected, tableMidWidth } from './ListView.module.scss'
import moment from 'moment'
import { faCaretDown } from '@fortawesome/free-solid-svg-icons/faCaretDown'
import { faCaretUp } from '@fortawesome/free-solid-svg-icons/faCaretUp'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

export default function ListView({ columns = {}, rows = [], colorRow = () => { } }) {
    const [sortColumn, setSortColumn] = useState("")
    const [ascending, setAscending] = useState(true)
    const [listItems, setListItems] = useState(rows)
    const [selectedAppointment, setSelectedAppointment] = useState(null)

    function sort(sortOrder) {
        return [...listItems].sort((listItem1, listItem2) => {
            if (sortColumn.includes("Date")) {
                const listItem1SortColumn = moment(listItem1[sortColumn]),
                    listItem2SortColumn = moment(listItem2[sortColumn])
                return compareTime(listItem1SortColumn, listItem2SortColumn, sortOrder)
            } else if (sortColumn.includes("Time")) {
                const listItem1SortColumn = moment(listItem1[sortColumn], 'h:mm a'),
                    listItem2SortColumn = moment(listItem2[sortColumn], 'h:mm a')

                return compareTime(listItem1SortColumn, listItem2SortColumn, sortOrder)
            } else {
                const listItem1SortColumn = listItem1[sortColumn] ? listItem1[sortColumn].toLowerCase() : "",
                    listItem2SortColumn = listItem2[sortColumn] ? listItem2[sortColumn].toLowerCase() : ""
                return compare(listItem1SortColumn, listItem2SortColumn, sortOrder)
            }
        })
    }

    function compareTime(time1, time2, sortOrder) {
        if (time1.isBefore(time2)) {
            return -1 * sortOrder
        } else if (time1.isAfter(time2)) {
            return 1 * sortOrder
        } else {
            return 0
        }
    }

    function compare(item1, item2, sortOrder) {
        if (item1 < item2) {
            return -1 * sortOrder
        } else if (item1 > item2) {
            return 1 * sortOrder
        } else {
            return 0
        }
    }


    useEffect(() => {
        if (ascending) {
            setListItems(sort(1))
        } else {
            setListItems(sort(-1))
        }
    }, [ascending, sortColumn])


    function renderSortIcon(column) {
        const icon = ascending ? faCaretUp : faCaretDown
        if (sortColumn === column) {
            return <FontAwesomeIcon data-testid="sortIcon" icon={icon} className={classNames(sortIcon)} onClick={() => setAscending(!ascending)} />
        }
        return null
    }

    if (listItems.length === 0) {
        return (
            <div className={classNames(noAppointmentsListView)}>
                No appointments found
            </div>)
    }
    return (
        <table className={classNames(appointmentListView)}>
            <thead>
                <tr>
                    {Object.keys(columns).map(column => {
                        let contentLength = ""
                        if (columns[column] === "mid") {
                            contentLength = tableMidWidth
                        }
                        return (
                            <th key={column} className={classNames(contentLength)} onClick={() => setSortColumn(column)}>
                                {column}
                                {renderSortIcon(column)}
                            </th>
                        );
                    })}
                </tr>
            </thead>
            <tbody>
                {listItems.map((listItem, index) => {
                    return (<tr key={index} className={selectedAppointment === listItem ? classNames(selected) : null} onClick={() => setSelectedAppointment(listItem)}>
                        {Object.keys(listItem).map((header, index) => {

                            let contentLength = ""
                            if (columns[header] === "mid") {
                                contentLength = tableMidWidth
                            }
                            return (<td key={index} className={classNames(contentLength, colorRow(listItem))}>{listItem[header]}</td>)
                        })}
                    </tr>)
                })}
            </tbody>
        </table>)
}