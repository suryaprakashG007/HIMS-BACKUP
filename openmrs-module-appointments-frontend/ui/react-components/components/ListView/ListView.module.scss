@import '../../variables.scss';

.appointmentListView {
    overflow-x: auto;
    overflow-y: hidden;
    height: calc(100vh - 230px);
    position: relative;
    border-collapse: collapse;
    th, td {
      font-size: 14px;
      padding: 8px;
      border: 1px solid #ddd;
      width: 90px;
      min-width: 90px;
      max-width: 90px;
      overflow: hidden;
      text-align: center;
      word-wrap: break-word;
      &.tableMidWidth {
        width: 180px;
        min-width: 180px;
        max-width: 180px;
      }
      &:last-child {
        width: 540px;
        min-width: 540px;
        max-width: 540px;
      }
    }
    thead {
      position: absolute;
      top: 0;
      left: 0;
      z-index: 1;
      th {
        font-weight: 900;
        color: #4c4c4c;
        background: #e2e1e1;
        border-color: #e2e1e1;
        white-space: nowrap;
        text-overflow: ellipsis;
      }
    }
    tbody {
      position: absolute;
      height: calc(100% - 51px);
      overflow-y: auto;
      top: 33px;
      tr {
        &:nth-child(even) {
          background: #F2F1F1;
        }
        &.selected {
          td {
            color: inherit;
            background: #298dd5;
            border-color: #298dd5;
          }
        }
      }
    }
  }
  
  .cancelledAppointment {
    background-color: $color-cancelled-appointment-color
  }

  .missedAppointment {
    background-color: $color-missed-appointment-color
  }

  .noAppointmentsListView {
    text-align: center;
  }

  .sortIcon {
    padding-left: 8px;
  }