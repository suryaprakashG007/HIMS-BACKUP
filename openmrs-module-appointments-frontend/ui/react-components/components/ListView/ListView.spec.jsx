import React from 'react'
import ListView from './ListView'
import { render, fireEvent } from '@testing-library/react'
import { missedAppointment, cancelledAppointment } from './ListView.module.scss'
import { toHaveClass } from '@testing-library/jest-dom/extend-expect'

describe('List View Component', () => {

    it('renders ListView with no appointments correctly', () => {
        const { asFragment } = render(<ListView />)

        expect(asFragment()).toMatchSnapshot()
    })

    const columns = {
        "Patient ID": "",
        "Patient Name": "mid",
        "Date": "",
        "Start Time": "",
        "End Time": "",
        "Provider": "mid",
        "Speciality": "",
        "Service": "mid",
        "Service Appointment Type": "mid",
        "Status": "",
        "Walk In": "",
        "Location": "mid",
        "Additional Info": "mid",
        "Notes": ""
    }

    const rows = [
        {
            "Patient ID": "GAN203012",
            "Patient Name": "Test Diabetes",
            "Date": "24 Mar 20",
            "Start Time": "10:00 am",
            "End Time": "10:30 am",
            "Provider": "",
            "Speciality": "Cardiology",
            "Service": "Cardiology",
            "Service Appointment Type": "",
            "Status": "Scheduled",
            "Walk In": "Yes",
            "Location": "OPD-1",
            "Additional Info": "",
            "Notes": "",
        },
        {
            "Patient ID": "GAN203011",
            "Patient Name": "Test Con",
            "Date": "24 Mar 20",
            "Start Time": "9:00 am",
            "End Time": "9:30 am",
            "Provider": "",
            "Speciality": "Cardiology",
            "Service": "Cardiology",
            "Service Appointment Type": "",
            "Status": "Scheduled",
            "Walk In": "Yes",
            "Location": "OPD-1",
            "Additional Info": "",
            "Notes": "Urgent",
        },
        {
            "Patient ID": "GAN203019",
            "Patient Name": "Test Con",
            "Date": "22 Mar 20",
            "Start Time": "11:00 am",
            "End Time": "11:30 am",
            "Provider": "",
            "Speciality": "Psychology",
            "Service": "Psychology",
            "Service Appointment Type": "",
            "Status": "Scheduled",
            "Walk In": "Yes",
            "Location": "General Ward",
            "Additional Info": "",
            "Notes": "Nothing important",
        }
    ]

    it('renders ListView with appointments correctly', () => {
        const { asFragment } = render(<ListView columns={columns} rows={rows} />)

        expect(asFragment()).toMatchSnapshot()
    })

    it('renders ListView with background colors based on status', () => {
        const coloredRows = [
            { ...rows[0], "Status": "Missed" },
            { ...rows[1], "Status": "Cancelled" },
            { ...rows[2], "Status": "Scheduled" }]
        const { asFragment } = render(<ListView columns={columns} rows={coloredRows} colorRow={(appointment) => {
            const statusToColor = { "Missed": missedAppointment, "Cancelled": cancelledAppointment }
            return statusToColor[appointment.Status]
        }} />)

        expect(asFragment()).toMatchSnapshot()
    })


    it('should sort records ascending wrt column and display icon on click of column header', () => {
        rows[0]["Patient Name"] = "Antony"
        rows[1]["Patient Name"] = "Zenia"
        rows[2]["Patient Name"] = "Ramesh"
        const { getByText, container } = render(<ListView columns={columns} rows={rows} />)

        fireEvent.click(getByText('Patient Name'))
        expect(getByText("Patient Name").childNodes[1]).toHaveClass("svg-inline--fa fa-caret-up fa-w-10 sortIcon")

        const tableBody = container.querySelector("tbody")
        expect(tableBody.children[0].childNodes[1].textContent).toBe("Antony")
        expect(tableBody.children[1].childNodes[1].textContent).toBe("Ramesh")
        expect(tableBody.children[2].childNodes[1].textContent).toBe("Zenia")
    })

    it('should sort records descending wrt column and display icon on click of column header', () => {
        rows[0]["Patient Name"] = "Antony"
        rows[1]["Patient Name"] = "Zenia"
        rows[2]["Patient Name"] = "Ramesh"
        const { getByText, getByTestId, container } = render(<ListView columns={columns} rows={rows} />)

        fireEvent.click(getByText('Patient Name'))
        fireEvent.click(getByTestId("sortIcon"))

        expect(getByText("Patient Name").childNodes[1]).toHaveClass("svg-inline--fa fa-caret-down fa-w-10 sortIcon")

        const tableBody = container.querySelector("tbody")
        expect(tableBody.children[0].childNodes[1].textContent).toBe("Zenia")
        expect(tableBody.children[1].childNodes[1].textContent).toBe("Ramesh")
        expect(tableBody.children[2].childNodes[1].textContent).toBe("Antony")
    })

    it('should sort records in descending if previous column based sorting was descending', () => {
        rows[0]["Patient ID"] = "GAN00012"
        rows[1]["Patient ID"] = "GAN00020"
        rows[2]["Patient ID"] = "GAN00001"
        const { getByText, getByTestId, container } = render(<ListView columns={columns} rows={rows} />)

        fireEvent.click(getByText('Patient Name'))
        fireEvent.click(getByTestId("sortIcon"))
        expect(getByText("Patient Name").childNodes[1]).toHaveClass("svg-inline--fa fa-caret-down fa-w-10 sortIcon")
        fireEvent.click(getByText('Patient ID'))

        expect(getByText("Patient Name").childNodes.length).toEqual(1)
        expect(getByText("Patient ID").childNodes[1]).toHaveClass("svg-inline--fa fa-caret-down fa-w-10 sortIcon")
        const tableBody = container.querySelector("tbody")
        expect(tableBody.children[0].childNodes[0].textContent).toBe("GAN00020")
        expect(tableBody.children[1].childNodes[0].textContent).toBe("GAN00012")
        expect(tableBody.children[2].childNodes[0].textContent).toBe("GAN00001")
    })

    it('should select an appointment if it is clicked and deselect if another is selected', () => {
        const { container } = render(<ListView columns={columns} rows={rows} />)
        const row1Appointment = container.querySelector("tbody").children[0]
        const row2Appointment = container.querySelector("tbody").children[1]

        fireEvent.click(row1Appointment)
        expect(row1Appointment).toHaveClass('selected')

        fireEvent.click(row2Appointment)
        expect(row2Appointment).toHaveClass('selected')
        expect(row1Appointment).not.toHaveClass('selected')
    })

})
