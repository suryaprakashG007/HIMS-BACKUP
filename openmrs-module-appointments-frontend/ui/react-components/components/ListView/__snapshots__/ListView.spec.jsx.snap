// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List View Component renders ListView with appointments correctly 1`] = `
<DocumentFragment>
  <table
    class="appointmentListView"
  >
    <thead>
      <tr>
        <th
          class=""
        >
          Patient ID
        </th>
        <th
          class="tableMidWidth"
        >
          Patient Name
        </th>
        <th
          class=""
        >
          Date
        </th>
        <th
          class=""
        >
          Start Time
        </th>
        <th
          class=""
        >
          End Time
        </th>
        <th
          class="tableMidWidth"
        >
          Provider
        </th>
        <th
          class=""
        >
          Speciality
        </th>
        <th
          class="tableMidWidth"
        >
          Service
        </th>
        <th
          class="tableMidWidth"
        >
          Service Appointment Type
        </th>
        <th
          class=""
        >
          Status
        </th>
        <th
          class=""
        >
          Walk In
        </th>
        <th
          class="tableMidWidth"
        >
          Location
        </th>
        <th
          class="tableMidWidth"
        >
          Additional Info
        </th>
        <th
          class=""
        >
          Notes
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class=""
        >
          GAN203012
        </td>
        <td
          class="tableMidWidth"
        >
          Test Diabetes
        </td>
        <td
          class=""
        >
          24 Mar 20
        </td>
        <td
          class=""
        >
          10:00 am
        </td>
        <td
          class=""
        >
          10:30 am
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth"
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Scheduled
        </td>
        <td
          class=""
        >
          Yes
        </td>
        <td
          class="tableMidWidth"
        >
          OPD-1
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        />
      </tr>
      <tr>
        <td
          class=""
        >
          GAN203011
        </td>
        <td
          class="tableMidWidth"
        >
          Test Con
        </td>
        <td
          class=""
        >
          24 Mar 20
        </td>
        <td
          class=""
        >
          9:00 am
        </td>
        <td
          class=""
        >
          9:30 am
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth"
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Scheduled
        </td>
        <td
          class=""
        >
          Yes
        </td>
        <td
          class="tableMidWidth"
        >
          OPD-1
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Urgent
        </td>
      </tr>
      <tr>
        <td
          class=""
        >
          GAN203019
        </td>
        <td
          class="tableMidWidth"
        >
          Test Con
        </td>
        <td
          class=""
        >
          22 Mar 20
        </td>
        <td
          class=""
        >
          11:00 am
        </td>
        <td
          class=""
        >
          11:30 am
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Psychology
        </td>
        <td
          class="tableMidWidth"
        >
          Psychology
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Scheduled
        </td>
        <td
          class=""
        >
          Yes
        </td>
        <td
          class="tableMidWidth"
        >
          General Ward
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Nothing important
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`List View Component renders ListView with background colors based on status 1`] = `
<DocumentFragment>
  <table
    class="appointmentListView"
  >
    <thead>
      <tr>
        <th
          class=""
        >
          Patient ID
        </th>
        <th
          class="tableMidWidth"
        >
          Patient Name
        </th>
        <th
          class=""
        >
          Date
        </th>
        <th
          class=""
        >
          Start Time
        </th>
        <th
          class=""
        >
          End Time
        </th>
        <th
          class="tableMidWidth"
        >
          Provider
        </th>
        <th
          class=""
        >
          Speciality
        </th>
        <th
          class="tableMidWidth"
        >
          Service
        </th>
        <th
          class="tableMidWidth"
        >
          Service Appointment Type
        </th>
        <th
          class=""
        >
          Status
        </th>
        <th
          class=""
        >
          Walk In
        </th>
        <th
          class="tableMidWidth"
        >
          Location
        </th>
        <th
          class="tableMidWidth"
        >
          Additional Info
        </th>
        <th
          class=""
        >
          Notes
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="missedAppointment"
        >
          GAN203012
        </td>
        <td
          class="tableMidWidth missedAppointment"
        >
          Test Diabetes
        </td>
        <td
          class="missedAppointment"
        >
          24 Mar 20
        </td>
        <td
          class="missedAppointment"
        >
          10:00 am
        </td>
        <td
          class="missedAppointment"
        >
          10:30 am
        </td>
        <td
          class="tableMidWidth missedAppointment"
        />
        <td
          class="missedAppointment"
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth missedAppointment"
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth missedAppointment"
        />
        <td
          class="missedAppointment"
        >
          Missed
        </td>
        <td
          class="missedAppointment"
        >
          Yes
        </td>
        <td
          class="tableMidWidth missedAppointment"
        >
          OPD-1
        </td>
        <td
          class="tableMidWidth missedAppointment"
        />
        <td
          class="missedAppointment"
        />
      </tr>
      <tr>
        <td
          class="cancelledAppointment"
        >
          GAN203011
        </td>
        <td
          class="tableMidWidth cancelledAppointment"
        >
          Test Con
        </td>
        <td
          class="cancelledAppointment"
        >
          24 Mar 20
        </td>
        <td
          class="cancelledAppointment"
        >
          9:00 am
        </td>
        <td
          class="cancelledAppointment"
        >
          9:30 am
        </td>
        <td
          class="tableMidWidth cancelledAppointment"
        />
        <td
          class="cancelledAppointment"
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth cancelledAppointment"
        >
          Cardiology
        </td>
        <td
          class="tableMidWidth cancelledAppointment"
        />
        <td
          class="cancelledAppointment"
        >
          Cancelled
        </td>
        <td
          class="cancelledAppointment"
        >
          Yes
        </td>
        <td
          class="tableMidWidth cancelledAppointment"
        >
          OPD-1
        </td>
        <td
          class="tableMidWidth cancelledAppointment"
        />
        <td
          class="cancelledAppointment"
        >
          Urgent
        </td>
      </tr>
      <tr>
        <td
          class=""
        >
          GAN203019
        </td>
        <td
          class="tableMidWidth"
        >
          Test Con
        </td>
        <td
          class=""
        >
          22 Mar 20
        </td>
        <td
          class=""
        >
          11:00 am
        </td>
        <td
          class=""
        >
          11:30 am
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Psychology
        </td>
        <td
          class="tableMidWidth"
        >
          Psychology
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Scheduled
        </td>
        <td
          class=""
        >
          Yes
        </td>
        <td
          class="tableMidWidth"
        >
          General Ward
        </td>
        <td
          class="tableMidWidth"
        />
        <td
          class=""
        >
          Nothing important
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`List View Component renders ListView with no appointments correctly 1`] = `
<DocumentFragment>
  <div
    class="noAppointmentsListView"
  >
    No appointments found
  </div>
</DocumentFragment>
`;
